File: medium/015.3sum/3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
  let hash = {}
  for (let i = 0; i < nums.length; i++) {
    if (nums[i] in hash) {
      hash[nums[i]] += 1
    }
    else {
      hash[nums[i]] = 1
    }
  }
  const keys = Object.keys(hash)
  const positive = keys.map(parseInt).filter(n => n > 0)
  const negative = keys.map(parseInt).filter(n => n < 0)
  console.log(keys.map(parseInt))
  // console.log(negative)
  if (positive.length === 0 || negative.length === 0 || nums.length < 3) {
    return []
  }
  if (hash[0] > 3) {
    ans = [[0, 0, 0]]
  }
  else {
    ans = []
  }
  for (let i = 0; i < negative.length; i++) {
    for (let j = 0; j < positive.length; j++) {
      const result = -(+negative[i] + (+positive[j]))
      if (result === negative[i] && hash[result] > 1) {
        ans.push([negative[i], negative[i], positive[j]])
      }
      else if (result === positive[j] && hash[result] > 1) {
        ans.push([negative[i], positive[j], positive[j]])
      }
      else if (result in hash && negative[i] < result < positive[j]) {
        ans.push([negative[i], result, positive[j]])
      }
    }
  }
  return ans
};

console.log(threeSum([-1, -2, -3, 4, 1, 3, 0, 3, -2, 1, -2, 2, -1, 1, -5, 4, -3]
))

File: medium/015.3sum/3sum.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        ans = []
        if not nums or nums[len(nums) - 1] < 0:
            return []
        for i in range(0, len(nums)-2):
            if nums[i] > 0:
                break
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            j, k = i + 1, len(nums) - 1
            while j < k and nums[k] >= 0:
                result = nums[i] + nums[j] + nums[k]
                if result == 0 and [nums[i], nums[j], nums[k]] not in ans:
                    ans.append([nums[i], nums[j], nums[k]])
                    k -= 1
                    while j < k-1 and nums[k] == nums[k+1]:
                        k -= 1
                    j += 1
                    while j + 1 < k and nums[j - 1] == nums[j]:
                        j += 1
                elif result > 0:
                    k -= 1
                    while j < k-1 and nums[k] == nums[k+1]:
                        k -= 1
                else:
                    j += 1
                    while j + 1 < k and nums[j - 1] == nums[j]:
                        j += 1
        return ans


if __name__ == "__main__":
    s = Solution()
    threeSum = s.threeSum

    print(threeSum([-1, 0, 1, 2, -1, -4]) == [[-1, -1, 2], [-1, 0, 1]])
    print(threeSum([0, 0, 0, 0]) == [[0, 0, 0]])
    print(threeSum([]) == [])
